Adopt more smart pointers in RenderElement

cdumez opened the pull request:
#### ea4f8e5ea638e3d82f8a87811aa8c9654cabf993
<pre>
Adopt more smart pointers in RenderElement
<a href="https://bugs.webkit.org/show_bug.cgi?id=263542">https://bugs.webkit.org/show_bug.cgi?id=263542</a>

Reviewed by Ryosuke Niwa.

Adopt more smart pointers in RenderElement.

I also made some changes to EventHandler so that we can store it in
a CheckedPtr / CheckedRef. I made it forward its checked pointer count
to its associated frame since it is owned by the frame. As extra safety,
I made EventHandler::m_frame be a CheckedRef&lt;Frame&gt; which required me
to update all uses of m_frame in this file. That said, the changes in
EventHandler are NOT meant to be exhaustive.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::protectedView const):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::incrementPtrCount const):
(WebCore::EventHandler::decrementPtrCount const):
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::checkedEventHandler):
(WebCore::LocalFrame::checkedEventHandler const):
(WebCore::Document::protectedView const): Deleted.
* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::createFor):
(WebCore::RenderElement::updateFillImages):
(WebCore::RenderElement::updateShapeImage):
(WebCore::RenderElement::repaintBeforeStyleChange):
(WebCore::RenderElement::initializeStyle):
(WebCore::RenderElement::setStyle):
(WebCore::RenderElement::didAttachChild):
(WebCore::RenderElement::attachRendererInternal):
(WebCore::RenderElement::detachRendererInternal):
(WebCore::findNextLayer):
(WebCore::layerNextSiblingRespectingTopLayer):
(WebCore::addLayers):
(WebCore::RenderElement::removeLayers):
(WebCore::RenderElement::moveLayers):
(WebCore::RenderElement::layerParent const):
(WebCore::RenderElement::propagateStyleToAnonymousChildren):
(WebCore::RenderElement::styleWillChange):
(WebCore::RenderElement::styleDidChange):
(WebCore::RenderElement::insertedIntoTree):
(WebCore::RenderElement::willBeRemovedFromTree):
(WebCore::RenderElement::clearSubtreeLayoutRootIfNeeded const):
(WebCore::RenderElement::willBeDestroyed):
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::protectedFrame const):
* Source/WebCore/rendering/RenderObjectInlines.h:
(WebCore::RenderObject::protectedDocument const):
* Source/WebCore/rendering/style/NinePieceImage.h:
(WebCore::NinePieceImage::protectedImage const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::protectedBackgroundLayers const):
(WebCore::RenderStyle::protectedMaskLayers const):
(WebCore::RenderStyle::protectedShapeOutside const):
* Source/WebCore/rendering/style/ShapeValue.h:
(WebCore::ShapeValue::protectedImage const):

Canonical link: <a href="https://commits.webkit.org/269712@main">https://commits.webkit.org/269712@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/73eae543c2aa88b7e9eab814701f04bcdebdd09d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23330 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1444 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24453 "Built successfully") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21552 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23601 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2975 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23886 "Built successfully") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23571 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1000 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20204 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26097 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/807 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21091 "Passed tests") | [💥 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27241 "An unexpected error occured. Recent messages:Failed to print configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21271 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21353 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25115 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/802 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18548 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/759 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1215 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2975 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1061 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -1706,7 +1700,7 @@ std::optional<Cursor> EventHandler::selectCursor(const HitTestResult& result, bo
 void EventHandler::startAutoHideCursorTimer()
 {
-    Page* page = m_frame.page();
+    Page* page = m_frame->page();
```
Why not use checkedPage here as well?

cdumez replied:
The changes to EventHandler are not meant to be exhaustive. This patch's focus is RenderElement. I just made some changes to EventHandler store it in a CheckedPtr/CheckedRef and changing m_frame ended up making the change large.

Looking at this particular function though, you can see the page is only used to query a setting. I figure this was safe so I didn't opportunistically use CheckedPtr for the page here.

rniwa commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -1728,11 +1722,11 @@ void EventHandler::cancelAutoHideCursorTimer()
 void EventHandler::autoHideCursorTimerFired()
 {
-    auto* view = m_frame.view();
+    auto* view = m_frame->view();
```
Use RefPtr?

cdumez replied:
It only called `view->isActive()` so it seemed excessive?

rniwa commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -3698,13 +3692,13 @@ bool EventHandler::isKeyEventAllowedInFullScreen(const PlatformKeyboardEvent& ke
 
-    if (m_frame.document())
-        savedLastHandledUserGestureTimestamp = m_frame.document()->lastHandledUserGestureTimestamp();
+    if (auto* document = frame->document())
```
Use RefPtr?

cdumez replied:
We only call a trivial getter on the document so this seemed excessive.

Also note that there are still plenty of potentially unsafe pointer usage in this file. The focus on this patch is RenderElement, not EventHandler.

rniwa commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -4308,10 +4302,10 @@ bool EventHandler::handleDrag(const MouseEventWithHitTestResults& event, CheckDr
     invalidateDataTransfer();
 
-    if (!m_frame.document())
+    if (!frame->document())
```
Maybe we can do:
```cpp
RefPtr document = frame->document();
if (document)
    return false;

dragState().dataTransfer = DataTransfer::createForDrag(*document);
```

rniwa commented on Source/WebCore/page/EventHandler.h:
```diff
@@ -136,6 +137,10 @@ class EventHandler {
     ~EventHandler();
 
+    // For CheckedPtr / CheckedRef use.
+    void incrementPtrCount() const;
```
I think we should just make EventHandler inherit from `CanMakeCheckedPtr` unless this is absolutely necessary for correctness, in which case the reason should be stated in the commit message.

cdumez replied:
I think this goes against @darinadler 's recommendation on my earlier PR?

The EventHandler is completely owned by the Frame via a UniqueRef so we don't need to increase the size of EventHandler unnecessarily. I also made EventHandler::m_frame be a CheckedRef<Frame> so that the lifetime safety is bi-directional.

I don't mind either way. I merely went this way to be consistent with my earlier PR where I think Darin suggested this (the review comments were not 100% clear to me so I may have misunderstood as well).

rniwa commented on Source/WebCore/page/mac/EventHandlerMac.mm:
```diff
@@ -734,7 +734,7 @@ static bool frameHasPlatformWidget(const LocalFrame& frame)
 {
     NSView *windowView = nil;
-    if (Page* page = m_frame.page())
+    if (Page* page = m_frame->page())
```
Use CheckedPtr?

rniwa commented on Source/WebCore/page/mac/EventHandlerMac.mm:
```diff
@@ -338,7 +338,7 @@ static bool findViewInSubviews(NSView *superview, NSView *target)
         return nil;
     }
-    auto* topFrameView = m_frame.view();
+    auto* topFrameView = m_frame->view();
```
Use RefPtr?

rniwa commented on Source/WebCore/page/mac/EventHandlerMac.mm:
```diff
@@ -518,7 +518,7 @@ static void selfRetainingNSScrollViewScrollWheel(NSScrollView *self, SEL selecto
 void EventHandler::mouseDown(NSEvent *event, NSEvent *correspondingPressureEvent)
 {
-    auto* v = m_frame.view();
+    auto* v = m_frame->view();
```
Use RefPtr?

rniwa commented on Source/WebCore/page/mac/EventHandlerMac.mm:
```diff
@@ -535,7 +535,7 @@ static void selfRetainingNSScrollViewScrollWheel(NSScrollView *self, SEL selecto
 void EventHandler::mouseDragged(NSEvent *event, NSEvent *correspondingPressureEvent)
 {
-    auto* v = m_frame.view();
+    auto* v = m_frame->view();
```
Use RefPtr?

rniwa commented on Source/WebCore/page/mac/EventHandlerMac.mm:
```diff
@@ -549,7 +549,7 @@ static void selfRetainingNSScrollViewScrollWheel(NSScrollView *self, SEL selecto
 void EventHandler::mouseUp(NSEvent *event, NSEvent *correspondingPressureEvent)
 {
-    auto* v = m_frame.view();
+    auto* v = m_frame->view();
```
Use RefPtr?

rniwa commented on Source/WebCore/page/mac/EventHandlerMac.mm:
```diff
@@ -586,7 +586,7 @@ static void selfRetainingNSScrollViewScrollWheel(NSScrollView *self, SEL selecto
 {
-    auto* view = m_frame.view();
+    auto* view = m_frame->view();
     if (!view)
```
Use RefPtr?

rniwa commented on Source/WebCore/page/mac/EventHandlerMac.mm:
```diff
@@ -864,11 +861,11 @@ static bool eventTargetIsPlatformWidget(Element* eventTarget)
         return;
 
-    auto* view = m_frame.view();
+    auto* view = m_frame->view();
```
Use RefPtr?

cdumez committed:
Adopt more smart pointers in RenderElement
https://bugs.webkit.org/show_bug.cgi?id=263542

Reviewed by Ryosuke Niwa.

Adopt more smart pointers in RenderElement.

I also made some changes to EventHandler so that we can store it in
a CheckedPtr / CheckedRef. I made it forward its checked pointer count
to its associated frame since it is owned by the frame. As extra safety,
I made EventHandler::m_frame be a CheckedRef<Frame> which required me
to update all uses of m_frame in this file. That said, the changes in
EventHandler are NOT meant to be exhaustive.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::protectedView const):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::incrementPtrCount const):
(WebCore::EventHandler::decrementPtrCount const):
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::checkedEventHandler):
(WebCore::LocalFrame::checkedEventHandler const):
(WebCore::Document::protectedView const): Deleted.
* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::createFor):
(WebCore::RenderElement::updateFillImages):
(WebCore::RenderElement::updateShapeImage):
(WebCore::RenderElement::repaintBeforeStyleChange):
(WebCore::RenderElement::initializeStyle):
(WebCore::RenderElement::setStyle):
(WebCore::RenderElement::didAttachChild):
(WebCore::RenderElement::attachRendererInternal):
(WebCore::RenderElement::detachRendererInternal):
(WebCore::findNextLayer):
(WebCore::layerNextSiblingRespectingTopLayer):
(WebCore::addLayers):
(WebCore::RenderElement::removeLayers):
(WebCore::RenderElement::moveLayers):
(WebCore::RenderElement::layerParent const):
(WebCore::RenderElement::propagateStyleToAnonymousChildren):
(WebCore::RenderElement::styleWillChange):
(WebCore::RenderElement::styleDidChange):
(WebCore::RenderElement::insertedIntoTree):
(WebCore::RenderElement::willBeRemovedFromTree):
(WebCore::RenderElement::clearSubtreeLayoutRootIfNeeded const):
(WebCore::RenderElement::willBeDestroyed):
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::protectedFrame const):
* Source/WebCore/rendering/RenderObjectInlines.h:
(WebCore::RenderObject::protectedDocument const):
* Source/WebCore/rendering/style/NinePieceImage.h:
(WebCore::NinePieceImage::protectedImage const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::protectedBackgroundLayers const):
(WebCore::RenderStyle::protectedMaskLayers const):
(WebCore::RenderStyle::protectedShapeOutside const):
* Source/WebCore/rendering/style/ShapeValue.h:
(WebCore::ShapeValue::protectedImage const):

Canonical link: https://commits.webkit.org/269712@main
